### ISSUE TITLE
Questions Regarding Model Evaluation and Data Processing

### ISSUE THREAD

**Issue by qqq202** [idea-iitd/NeuroCUT#3]
Dear Author,

Thank you for quickly uploading the `finetune.py` file. I have closed the previous issue and now face several new questions as I am a beginner in this field. I have some doubts while learning the paper and code, and I would appreciate your guidance if my understanding is incorrect.

1. **Evaluation under 4 Criteria**:
   According to Tables 3-6 in the paper, the model is evaluated using 4 criteria: Normalized Cut, Min Cut, Sparsest Cut, and Balanced Cut. However, the `makefile` only provides options for `normalised`, `kmin`, and `sparsest` cuts, but not for `Balanced Cut`. How can I evaluate the model under all 4 criteria as shown in the paper?

2. **Experimenting with Other Methods**:
   I would like to conduct experiments using other methods under the same 4 criteria in your framework. I noticed that the `getNormalisedCutValue` function supports values such as `mincutpool` and `GAP` for `cuttype`. Could you provide guidance on how to set up and run these experiments?

3. **Additional Attributes in `graph_stats.txt`**:
   In the `graph_stats.txt` file provided in the `data` folder for the `cora_lc` dataset, there are additional attributes: `hMetis_norm_cut`, `spectral_norm_cut`, and `spectral_norm_cut_35`, besides `num_nodes`, `num_edges`, and `num_cuts`. When I follow the `README` to process the Cora dataset from `dataset = Planetoid(name='Cora', transform=T.NormalizeFeatures())`, the `graph_stats.txt` file only contains `num_nodes`, `num_edges`, and `num_cuts`. How can I obtain the additional attributes (`hMetis_norm_cut`, `spectral_norm_cut`, `spectral_norm_cut_35`) for my processed dataset?

Thank you in advance for your help and support.

Best regards,
qqq202

**Comment by sahilm1992** [comment]
Hello,
Thanks for your interest in our work. We will get back to you soon on this.

Regards
Sahil

**Comment by qqq202** [comment]
Hi Sahil,
﻿
Thank you for your quick response! I'm really looking forward to your guidance.
﻿
Best regards,
qqq202

**Comment by rishi1001** [comment]
Thank you for your interest in our work! Below are the answers to your questions:

1. To test for a `balanced cut`, please use `cuttype = "GAP_1"` in the code. We refer to it as `GAP_1` in the paper, as balanced cut was introduced in the GAP method ([reference](https://arxiv.org/abs/1903.00614)).

2. Yes, our code supports various other objective functions (cuttypes). Simply set the `cuttype` argument to the desired objective function to test different criteria. Additionally, you can create custom objectives by modifying the `src/model_utils.py` file.

3. The only required statistics to run the code are `num_nodes`, `num_edges`, and `num_cuts`. The other statistics are optional baseline metrics, which can be generated using external methods such as hMetis ([link](http://vlsicad.eecs.umich.edu/BK/Slots/cache/www-users.cs.umn.edu/~karypis/metis/hmetis/)) or spectral clustering. These baselines are not included in our code.

I hope this answers all your questions.

**Comment by qqq202** [comment]
Thank you for your detailed explanation. It has helped me understand how to evaluate NeuroCUT according to the 4 criteria in the paper and other standards. I would greatly appreciate further guidance on how to use these 4 criteria to evaluate other methods.

For example, for the GAP method ([paper link](https://arxiv.org/abs/1903.00614)), after implementing its network structure, should I train it using the objective function from the original GAP paper (i.e., equation (9), which you have already implemented as `cuttype = "GAP_1"` in the `getNormalisedCutValue` function, thank you very much)?

My understanding is that during training, I should follow the specific training method and objective function of the method being compared. For evaluation, I plan to set the `cuttype` in the `calculate_stats` function of `inference.py` to `normalised`, `kmin`, `sparsest`, and `GAP_1` respectively, to assess the method's performance on the test set. Could you please confirm if my understanding is correct?

Thank you again for your time and assistance. Your guidance is invaluable to me.

Best regards,
qqq202

**Comment by rishi1001** [comment]
Yes, your understanding is correct. To evaluate GAP, you need to train it on the original objective function from that paper. Once you have the cuts from GAP, you can use `getNormalisedCutValue` in `src/model_utils.py` to get the different cut values.

**Comment by qqq202** [comment]
### Training and Evaluation Issues with NeuroCUT

Hello,

Thank you so much for confirming my understanding. Your guidance has been incredibly helpful, and I have successfully used the four criteria to evaluate other models. However, I am currently facing some challenges with the evaluation results when training NeuroCUT. Here are the commands I used:

**Training Command:**
```bash
python main.py --cuttype normalised --update_reward last --pool mean --wandb false --num_perturbation 1 --num_run 5 --initial_order random --initial_type kmeans_Linf --train cora_lc --gpu 0 --embedding given_lipchitz --anchors 35 --epochs 3 --model_name graphsage --num_layers 2 --need_training true --norm percentile --modelScore ModelLinkPred --iter_update_params 2 --isAssigned_feature false --scoring_func mlp --node_select true --node_select_heuristic diff_max_balanced --hops 2 --gamma 0.99
```

**Evaluation Command:**
```bash
python inference.py --num_perturbation_inference 1000 --update_reward last --pool mean --wandb false --num_perturbation 1 --num_run 5 --initial_order random --initial_type kmeans_Linf --train cora_lc --gpu 0 --embedding given_lipchitz --anchors 35 --epochs 3 --model_name graphsage --num_layers 2 --need_training true --norm percentile --modelScore ModelLinkPred --iter_update_params 2 --isAssigned_feature false --scoring_func mlp --node_select true --node_select_heuristic diff_max_balanced --hops 2
```

Despite using these commands, the evaluation results for NeuroCUT are not as expected. I would greatly appreciate any suggestions or insights you could provide to improve the training and evaluation process.

Thank you very much for your time and assistance. Your support is invaluable to me!

Best regards,

qqq202

**Comment by rishi1001** [comment]
Hello,

To test the code, you only need to run the main file. The inference file is provided solely for evaluating an already saved model.

I’m attaching the `yaml` configuration file for the experiment you’re attempting to run. Please refer to this file for setting up your experiment.

---

### YAML Configuration Details:

```yaml
wandb_version: 1

_wandb:
  desc: null
  value:
    cli_version: 0.14.0
    framework: torch
    is_jupyter_run: false
    is_kaggle_kernel: false
    python_version: 3.9.16

alpha_reward: 
  value: 0.0
anchors: 
  value: 35
beta: 
  value: 100
cuttype: 
  value: normalised
embedding: 
  value: given_lipchitz
epochs: 
  value: 40
gamma: 
  value: 0.99
gpu: 
  value: 3
hops: 
  value: 2
initial_order: 
  value: random
initial_type: 
  value: kmeans_Linf
isAssigned_feature: 
  value: 'false'
iter_update_params: 
  value: 2
modelPhase2: 
  value: ModelLocalLinkPred
model_name: 
  value: graphsage
need_training: 
  value: 'true'
node_select: 
  value: 'true'
node_select_heuristic: 
  value: diff_max_balanced
norm: 
  value: percentile
num_layers: 
  value: 2
num_perturbation: 
  value: 2.0
num_run: 
  value: 1
pool: 
  value: mean
scoring_func: 
  value: mlp
train: 
  value: cora_lc
update_reward: 
  value: last
wandb: 
  value: 'true'
```

---

Please refer to this configuration when running the experiment. The cut values you obtain from the main file should match our expected results.

Best regards,  
Rishi

**Comment by qqq202** [comment]
Hello Rishi,
﻿
Thank you very much for your detailed response and the YAML configuration file. I followed the provided settings and ran the experiment using the main file. The performance I obtained is now very close to the results reported in your paper.
﻿
Thank you again for your guidance and support. Your help has been invaluable to me!
﻿
Best regards,
﻿
qqq202
